File: src/Phoenix/Communication/WinSock.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Runtime.InteropServices;

namespace Phoenix.Communication
{
    public static class WinSock
    {
        private static ComInterop.IWinSock winSock = null;
        private static int threadId = -1;

        internal static void Initialize(ComInterop.IWinSock winSock)
        {
            WinSock.winSock = winSock;
            threadId = Thread.CurrentThread.ManagedThreadId;
        }

        public static int recv(long socket, byte[] data, int len, int flags)
        {
            if (threadId != Thread.CurrentThread.ManagedThreadId)
                throw new InvalidOperationException("Called from invalid thread.");

            return winSock.recv(socket, data, len, flags);
        }

        public static int send(long socket, byte[] data, int len, int flags)
        {
            if (threadId != Thread.CurrentThread.ManagedThreadId)
                throw new InvalidOperationException("Called from invalid thread.");

            return winSock.send(socket, data, len, flags);
        }

        public static int closesocket(long socket)
        {
            return winSock.closesocket(socket);
        }

        public static void SimulateCursorPos(int x, int y)
        {
            winSock.SimulateCursorPos(x, y);
        }

        public static void OverrideCursor()
        {
            winSock.OverrideCursor();
        }

        public static void ReleaseCursorOverride()
        {
            winSock.ReleaseCursorOverride();
        }
    }
}

File: src/Native/WinSock.cpp
#include "StdAfx.h"
#include "ComInterop.h"
#include "WinSock.h"


UNKNOWN_BEGIN(WinSock)
	UNKNOWN_QUERYINTERFACE(__uuidof(IWinSock), IWinSock)
UNKNOWN_END()

WinSock::WinSock()
	: UNKNOWN_INIT()
{
}

WinSock::~WinSock()
{
}


HRESULT __stdcall WinSock::recv (/*[in]*/ __int64 s, /*[out]*/ unsigned char * buf,
        /*[in]*/ long len, /*[in]*/ long flags, /*[out,retval]*/ long * pRetVal )
{
	*pRetVal = ::recv((SOCKET)s, (char*)buf, len, flags);
	return S_OK;
}

HRESULT __stdcall WinSock::send (/*[in]*/ __int64 s, /*[in]*/ unsigned char * buf,
        /*[in]*/ long len, /*[in]*/ long flags, /*[out,retval]*/ long * pRetVal )
{
	*pRetVal = ::send((SOCKET)s, (char*)buf, len, flags);
	return S_OK;
}

HRESULT __stdcall WinSock::closesocket (
        /*[in]*/ __int64 s,
        /*[out,retval]*/ long * pRetVal )
{
	*pRetVal = ::closesocket((SOCKET)s);
	return S_OK;
}


void SimulateCursorPos(int x, int y);
void OverrideCursor();
void ReleaseCursorOverride();

HRESULT __stdcall WinSock::SimulateCursorPos(long x, long y, /*[out,retval]*/ long * pRetVal)
{
    ::SimulateCursorPos(x, y);
    return 0;
}

HRESULT __stdcall WinSock::OverrideCursor(/*[out,retval]*/ long * pRetVal)
{
    ::OverrideCursor();
    return 0;
}

HRESULT __stdcall WinSock::ReleaseCursorOverride(/*[out,retval]*/ long * pRetVal)
{
    ::ReleaseCursorOverride();
    return 0;
}
